Add support for string keys in array unpacking

nikic opened the pull request:
RFC: https://wiki.php.net/rfc/array_unpacking_string_keys

TysonAndre commented on Zend/zend_vm_execute.h:
```diff
@@ -2560,29 +2560,34 @@ static ZEND_OPCODE_HANDLER_RET ZEND_FASTCALL ZEND_ADD_ARRAY_UNPACK_SPEC_HANDLER(
-					zend_cannot_add_element();
-					zval_ptr_dtor_nogc(val);
+				if (Z_TYPE(key) == IS_STRING) {
+					zend_hash_update(result_ht, Z_STR(key), val);
+					zval_ptr_dtor_str(&key);
```
1. Should this handle `[...new ArrayObject(['key' => ThrowsOnDestruct]), ...new CustomTraversable()]` to avoid an call to CustomTraversable::valid() in the unlikely case that an exception was thrown when the overwritten value's destructor was called?
   (i.e. before `iter->funcs->valid(iter) == SUCCESS` is called again)

   (e.g. in case CustomTraversable modifies state, performs db operations or makes network service calls)
2. nit: the original works correctly, but I wonder if the below snippet would be a tiny bit more efficient, because the C optimizer can't infer that Z_STR(key) doesn't change after the call to zend_hash_update(), which may call destructors which modify zval, etc.

Same for the other calls to zval_ptr_dtor_str() in this PR - I mostly see zval_ptr_dtor_str in this file places that the compiler can infer don't have side effects, such as zend_fast_equal_strings

```suggestion
					zend_string *key_str = Z_STR(key);                    
					zend_hash_update(result_ht, key_str, val);
					zend_string_release_ex(key_str, 0);

					if (UNEXPECTED(EG(exception) != NULL)) {
						break;
					}
```

TysonAndre reviewed (commented):
LGTM other than the question about checking for an exception before calling `Traversable->valid()`

TysonAndre commented on Zend/zend_vm_execute.h:
```diff
@@ -2560,29 +2560,34 @@ static ZEND_OPCODE_HANDLER_RET ZEND_FASTCALL ZEND_ADD_ARRAY_UNPACK_SPEC_HANDLER(
+					zval_ptr_dtor_str(&key);
+				} else {
+					if (!zend_hash_next_index_insert(result_ht, val)) {
+						zend_cannot_add_element();
+						zval_ptr_dtor_nogc(val);
```
NYC: why does the original code have no `break;` here if it's already throwing `Cannot add element to the array as the next element is already occupied`

TysonAndre reviewed (commented):
Could you add a test of the expected behavior for keys that are stringified integers, and clarify how this edge case is handled in the RFC?
Personally, I'd prefer using zend_symtable_update instead of zend_hash_update for Traversables (i.e. convert strings to integers like regular array insertion does)

```php
function my_generator() { yield '0' => 'test'; }
$x = [...my_generator()];
var_dump($x); // should this have an integer key 0 or a string key '0'?
```

To compare, iterator_to_array() currently converts strings to integers

```php
php > function my_generator() { yield '0' => 'test'; }
php > var_export(iterator_to_array(my_generator()));
array (
  0 => 'test',
)
php > foreach (my_generator() as $key => $_) { var_dump($key); }
string(1) "0"
```

nikic commented:
> Could you add a test of the expected behavior for keys that are stringified integers, and clarify how this edge case is handled in the RFC?
> Personally, I'd prefer using zend_symtable_update instead of zend_hash_update for Traversables (i.e. convert strings to integers like regular array insertion does)
> 
> ```
> function my_generator() { yield '0' => 'test'; }
> $x = [...my_generator()];
> var_dump($x); // should this have an integer key 0 or a string key '0'?
> ```

Nice catch! This definitely can't have a string key `'0'`, because that would violate the array key canonicalization. However, it's not immediately obvious whether this should result in a key `0` being set, or it being treated as an original integer key (and appended with renumbering). I think the latter behavior is probably correct...

TysonAndre commented:
>  I think the latter behavior is probably correct...

With the latter behavior, [...$a, ...$b] would behave similarly to `[...iterator_to_array($a, true), ...iterator_to_array($b, true)]`

With the former behavior, gaps in numbers added by `...` or starting to allow overwriting existing **numeric** keys in `...` would be unexpected

TysonAndre commented:
The remaining changes to opcache look straightforward


For ext/opcache/Optimizer/zend_inference.c , I think this should replace MAY_BE_ARRAY_KEY_LONG with `MAY_BE_ARRAY_KEY_ANY` or `(t1&MAY_BE_ARRAY_KEY_ANY)` 
I believe the current code would lead to improper JIT optimizations and dead code elimination for key types, and is straightforward to fix - may be worth an extra unit test.

```c
// original code
		case ZEND_ADD_ARRAY_UNPACK:
			tmp = ssa_var_info[ssa_op->result_use].type;
			ZEND_ASSERT(tmp & MAY_BE_ARRAY);
			/* Ignore string keys as they will throw. */
			if (t1 & MAY_BE_ARRAY_KEY_LONG) {
				tmp |= MAY_BE_ARRAY_KEY_LONG | (t1 & (MAY_BE_ARRAY_OF_ANY|MAY_BE_ARRAY_OF_REF));
			}
			if (t1 & MAY_BE_OBJECT) {
				tmp |= MAY_BE_ARRAY_KEY_LONG | MAY_BE_ARRAY_OF_ANY;
			}
			UPDATE_SSA_TYPE(tmp, ssa_op->result_def);
			break;
```

ext/opcache/Optimizer/sccp.c seems like the ct_eval_add_array_unpack can also be changed to permit string `key` (also with zend_symtable_update() to handle stringified integers)

```c
// original
static inline int ct_eval_add_array_unpack(zval *result, zval *array) {
	zend_string *key;
	zval *value;
	if (Z_TYPE_P(array) != IS_ARRAY) {
		return FAILURE;
	}

	SEPARATE_ARRAY(result);
	ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(array), key, value) {
		if (key) {
			return FAILURE;
		}
		value = zend_hash_next_index_insert(Z_ARR_P(result), value);
		if (!value) {
			return FAILURE;
		}
		Z_TRY_ADDREF_P(value);
	} ZEND_HASH_FOREACH_END();
```

nikic commented:
I've added handling for integral string keys and updated the RFC accordingly.

Possibly we should also use the same handling for argument unpacking (i.e. treat it as a positional argument rather than a certainly not existing argument name.)
